README.md
# LatLong provides a lightweight library for common latitude and longitude calculation.

This library supports both, the "Haversine" and the "Vincenty" algorithm.

"Haversine" is a bit faster but "Vincenty" is far more accurate!
 
<p align="center"> 
    <img alt="LatLong" src="https://github.com/MikeMitterer/dart-latlong/raw/master/doc/images/latlong.jpg"> 
</p>

[Catmull-Rom algorithm](http://hawkesy.blogspot.co.at/2010/05/catmull-rom-spline-curve-implementation.html) is used for smoothing out the path.
 
## Basic usage 

### Distance
```dart
    final Distance distance = new Distance();
    
    // km = 423
    final int km = distance.as(LengthUnit.Kilometer,
     new LatLng(52.518611,13.408056),new LatLng(51.519475,7.46694444));
    
    // meter = 422591.551
    final int meter = distance(
        new LatLng(52.518611,13.408056),
        new LatLng(51.519475,7.46694444)
        );

```

## Offset
```dart
    final Distance distance = const Distance();
    final num distanceInMeter = (EARTH_RADIUS * math.PI / 4).round();
    
    final p1 = new LatLng(0.0, 0.0);
    final p2 = distance.offset(p1, distanceInMeter, 180);
    
    // LatLng(latitude:-45.219848, longitude:0.0)
    print(p2.round());
    
    // 45° 13' 11.45" S, 0° 0' 0.00" O
    print(p2.toSexagesimal());
            
```

## Path smoothing
```dart
    // zigzag is a list of coordinates
    final Path path = new Path.from(zigzag);
    
    // Result is below
    final Path steps = path.equalize(8,smoothPath: true);
```
<p align="center"> 
    <img alt="Smooth path" src="https://github.com/MikeMitterer/dart-latlong/raw/master/doc/images/smooth-path.jpg">
</p>

## Features and bugs
Please file feature requests and bugs at the [issue tracker](https://github.com/MikeMitterer/dart-latlong/issues).

## License

    Copyright 2015 Michael Mitterer (office@mikemitterer.at),
    IT-Consulting and Development Limited, Austrian Branch

    Licensed under the Apache License, Version 2.0 (the "License");
    you may not use this file except in compliance with the License.
    You may obtain a copy of the License at

       http://www.apache.org/licenses/LICENSE-2.0

    Unless required by applicable law or agreed to in writing,
    software distributed under the License is distributed on an
    "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
    either express or implied. See the License for the specific language
    governing permissions and limitations under the License.


If this plugin is helpful for you - please [(Circle)](http://gplus.mikemitterer.at/) me
or **star** this repo here on GitHub


CHANGELOG.md
# Change Log for latlong
Lightweight library for common latitude and longitude calculation

## [Unreleased](http://github.com/mikemitterer/dart-latlong/compare/v0.6...HEAD)

### Refactor
* Dart 2 ready [9907d75](https://github.com/mikemitterer/dart-latlong/commit/9907d75acb720b0840f50e0c984c2e03db2764eb)

## [v0.5](http://github.com/mikemitterer/dart-latlong/compare/v0.4...v0.5) - 2018-06-06

### Refactor
* Dependency upgrade [b1a5152](https://github.com/mikemitterer/dart-latlong/commit/b1a5152934fef2425755d718ce232dfb7ee254bd)

## [v0.3.4](http://github.com/mikemitterer/dart-latlong/compare/v0.3.3...v0.3.4) - 2016-11-16

### Docs
* Strong-Mode support [86b858c](https://github.com/mikemitterer/dart-latlong/commit/86b858c8554bf9fcdda56ab33c247f34d8b17616)

## [v0.3.3](http://github.com/mikemitterer/dart-latlong/compare/v0.3.2...v0.3.3) - 2016-04-18

### Docs
* Better documentation [e9d453d](https://github.com/mikemitterer/dart-latlong/commit/e9d453d2b25ef8c8694b317882cf44ba50b5c028)

## [v0.3.2](http://github.com/mikemitterer/dart-latlong/compare/v0.3.1...v0.3.2) - 2016-02-04

### Feature
* Path.equalize excepts now a num instead of int [108431b](https://github.com/mikemitterer/dart-latlong/commit/108431ba9bcf326f4b551ea090c240be255315e6)

## [v0.3.1](http://github.com/mikemitterer/dart-latlong/compare/v0.3.0...v0.3.1) - 2016-02-03

### Feature
* Made Path more generic [ec42c04](https://github.com/mikemitterer/dart-latlong/commit/ec42c042cf9a7ec1d82e68a9458063713ec2b18b)

## [v0.3.0](http://github.com/mikemitterer/dart-latlong/compare/v0.2.1...v0.3.0) - 2016-01-29

### Feature
* CatmullRom added for Path-Finding [ebeeb2c](https://github.com/mikemitterer/dart-latlong/commit/ebeeb2c36e8c1368bda2e882426298e3f22c06b5)
* Algorithms got their own class [f8a661f](https://github.com/mikemitterer/dart-latlong/commit/f8a661fa80485853f2fafe774b52fd417dbb174f)

### Docs
* Improved Readme [f030f6c](https://github.com/mikemitterer/dart-latlong/commit/f030f6cb89cb71d4ae9d29a3b7ffa13b7ea2bac8)

## [v0.2.0](http://github.com/mikemitterer/dart-latlong/compare/v0.1.2...v0.2.0) - 2016-01-26

### Feature
* Circle#isInside Polygon#center [cce0a4d](https://github.com/mikemitterer/dart-latlong/commit/cce0a4da93f03977a4545c6470c85b5d5243637e)

## [v0.1.1](http://github.com/mikemitterer/dart-latlong/compare/v0.1.0...v0.1.1) - 2016-01-25

### Docs
* Basic usage in README [218116b](https://github.com/mikemitterer/dart-latlong/commit/218116bcab24af05d6a793ef4f4c3e32b96202e4)


This CHANGELOG.md was generated with [**Changelog for Dart**](https://pub.dartlang.org/packages/changelog)


lib/latlong/Distance.dart
/*
 * Copyright (c) 2016, Michael Mitterer (office@mikemitterer.at),
 * IT-Consulting and Development Limited.
 * 
 * All Rights Reserved.
 * 
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * 
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

part of latlong;

/// Calculates the distance between points.
///
/// Default algorithm is [distanceWithVincenty], default radius is [EARTH_RADIUS]
///
///      final Distance distance = new Distance();
///
///      // km = 423
///      final int km = distance.as(LengthUnit.Kilometer,
///         new LatLng(52.518611,13.408056),new LatLng(51.519475,7.46694444));
///
///      // meter = 422592
///      final int meter = distance(new LatLng(52.518611,13.408056),new LatLng(51.519475,7.46694444));
///
class Distance implements DistanceCalculator {
    // final Logger _logger = new Logger('latlong.Distance');

    final double _radius;
    final _roundResult;
    final DistanceCalculator _calculator;

    const Distance({ final bool roundResult: true, final DistanceCalculator calculator: const Vincenty() })
        : _radius = EARTH_RADIUS, _roundResult = roundResult, _calculator = calculator;

    Distance.withRadius(final double radius,
        { final bool roundResult: true, final DistanceCalculator calculator: const Vincenty()})
        : _radius = radius, _roundResult = roundResult, _calculator = calculator {

        Validate.isTrue(radius > 0, "Radius must be greater than 0 but was $radius");
    }

    double get radius => _radius;

    /// Returns either [Haversine] oder [Vincenty] calculator
    ///
    ///     final Distance distance = const DistanceHaversine();
    ///     final Circle circle = new Circle(base, 1000.0,calculator: distance.calculator);
    ///
    DistanceCalculator get calculator => _calculator;

    /// Shortcut for [distance]
    num call(final LatLng p1, final LatLng p2) {
        return distance(p1,p2);
    }

    /// Converts the distance to the given [LengthUnit]
    ///
    ///     final int km = distance.as(LengthUnit.Kilometer,
    ///         new LatLng(52.518611,13.408056),new LatLng(51.519475,7.46694444));
    ///
    num as(final LengthUnit unit,final LatLng p1, final LatLng p2) {
        final double dist = _calculator.distance(p1,p2);
        return _round(LengthUnit.Meter.to(unit,dist));
    }

    /// Computes the distance between two points.
    ///
    /// The function uses the [DistanceAlgorithm] specified in the CTOR
    @override
    num distance(final LatLng p1, final LatLng p2)
        => _round(_calculator.distance(p1,p2));

    /// Returns the great circle bearing (direction) in degrees to the next point ([p2])
    ///
    /// Find out about the difference between rhumb line and
    /// great circle bearing on [Wikipedia](http://en.wikipedia.org/wiki/Rhumb_line#General_and_mathematical_description).
    ///
    ///     final Distance distance = const Distance();
    ///
    ///     final LatLng p1 = new LatLng(0.0, 0.0);
    ///     final LatLng p2 = new LatLng(-90.0, 0.0);
    ///
    ///     expect(distance.direction(p1, p2), equals(180));
    double bearing(final LatLng p1, final LatLng p2) {
        final diffLongitude = p2.longitudeInRad - p1.longitudeInRad;

        final y = math.sin(diffLongitude);
        final x = math.cos(p1.latitudeInRad) * math.tan(p2.latitudeInRad) -
            math.sin(p1.latitudeInRad) * math.cos(diffLongitude);

        return radianToDeg(math.atan2(y, x));
    }

    /// Returns a destination point based on the given [distance] and [bearing]
    ///
    /// Given a [from] (start) point, initial [bearing], and [distance],
    /// this will calculate the destination point and
    /// final bearing travelling along a (shortest distance) great circle arc.
    ///
    ///     final Distance distance = const Distance();
    ///
    ///     final num distanceInMeter = (EARTH_RADIUS * math.PI / 4).round();
    ///
    ///     final p1 = new LatLng(0.0, 0.0);
    ///     final p2 = distance.offset(p1, distanceInMeter, 180);
    ///
    /// Bearing: Left - 270°, right - 90°, up - 0°, down - 180°
    @override
    LatLng offset(final LatLng from,final num distanceInMeter,final num bearing)
        => _calculator.offset(from,distanceInMeter.toDouble(),bearing.toDouble());

    //- private -----------------------------------------------------------------------------------

    double _round(final double value) => (_roundResult ? value.round().toDouble() : value);
}

/// Shortcut for
///     final Distance distance = const Distance(calculator: const Vincenty());
///
class DistanceVincenty extends Distance {

    const DistanceVincenty({ final bool roundResult: true })
        : super(roundResult: roundResult, calculator: const Vincenty());

    DistanceVincenty.withRadius(final double radius,
        { final bool roundResult: true })
        : super.withRadius(radius, roundResult: roundResult, calculator: const Vincenty()) {

        Validate.isTrue(radius > 0, "Radius must be greater than 0 but was $radius");
    }
}

/// Shortcut for
///     final Distance distance = const Distance(calculator: const Haversine());
///
class DistanceHaversine extends Distance {

    const DistanceHaversine({ final bool roundResult: true })
        : super(roundResult: roundResult, calculator: const Haversine());

    DistanceHaversine.withRadius(final double radius,
        { final bool roundResult: true })
        : super.withRadius(radius, roundResult: roundResult, calculator: const Haversine()) {

        Validate.isTrue(radius > 0, "Radius must be greater than 0 but was $radius");
    }
}







